### PR TITLE
Implement scaling for parameter vector of optimization.

### LINKED ISSUE #13: Add scaling of the parameter vector
Scale the parameter vector by the inverse of the start parameters in reparametrize_to_internal and undo the scaling in reparametrize_from_internal.

### PR DESCRIPTION
**Pull request by tobiasraabe**
Closes #13.

This PR implements two scaling features for the parameter vector of the optimization. Parameters can either be scaled by starting values or gradients.

### Todo

- [x] Add a little bit of documentation.

### CODE REVIEW

**Comment by tobiasraabe** [comment]
@janos There is an issue documented starting in ``internal_func`` in differentiation. It is because for scaling I have to set up an internal criterion function and pass it to gradient. There it gets wrapped again.

**Comment by janosg** [comment]
Whether to take derivatives for standard errors on external or internal params was discussed extensively in #57 and we decided that for now we take the derivatives on on the external params. 

Our derivative functions should work for functions that take numpy arrays as well. If they currently don't we have to rewrite `_create_internal_func`. The double wrapping is not very nice but the speed penalty is probably negligible. 

For the scaling we should probably use faster but less precise derivatives. So `extrapolation=False` and maybe even `method="forward"`. We should also use a larger stepsize. This can currently not be specified but I'll change it in #59.

File: estimagic/optimization/reparametrize.py [hunk 1]
```diff
@@ -420,3 +438,89 @@ def _fixed_from_internal(params_subset, constr):
     if value is not None:
         res["value"] = value
     return res["value"]
+
+
+def _rescale_to_internal(
```
Comment (janosg) [inline estimagic/optimization/reparametrize.py]: This should be broken up into two functions: One that calculates the scaling factor and one that does the rescaling based on the scaling factor. Then only the actual rescaling should be called in `reparametrize_to_internal`. The calculation of the scaling factor should be done in minimize and the resulting scaling factor should be part of the optimization output. 

My problem with the current implementation is that a call `reparametrize_to_internal` with a value that is different from the start values (for example if we want to calculate a gradient on the internal function), will trigger a re-calculation of the scaling factor. This not only slow but probably also not what the user wants. There should be at most one scaling factor per optimization.

**Comment by tobiasraabe** [comment]
@janosg I implemented the scaling like you suggested. The linting catches some unused arguments. You can go through the list or I can disable it.

**Comment by janosg** [comment]
@peisenha  I requested a review from you because you have implemented precondotioning for an optimizer before.

**Pull request merged**